homeassistant/components/cover/ryobi_gdo.py

```diff
@@ -0,0 +1,285 @@
+"""
+import logging
+import json
+import socket
+from datetime import timedelta
+import voluptuous as vol
+import requests
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.cover import (
+    CoverDevice, PLATFORM_SCHEMA, SUPPORT_OPEN, SUPPORT_CLOSE)
+from homeassistant.helpers.event import track_utc_time_change
+from homeassistant.const import (
+    CONF_DEVICE, CONF_USERNAME, CONF_PASSWORD,
+    STATE_UNKNOWN, STATE_CLOSED, STATE_OPEN, CONF_COVERS)
+
+REQUIREMENTS = ['websocket-client==0.37.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+COVER_SCHEMA = vol.Schema({
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_COVERS): vol.Schema({cv.slug: COVER_SCHEMA}),
+})
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+RYOBI_API_KEY_URL = "https://tti.tiwiconnect.com/api/login"
+RYOBI_DEVICES_URL = "https://tti.tiwiconnect.com/api/devices"
+RYOBI_WS_URL = "wss://tti.tiwiconnect.com/api/wsrpc"
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Ryobi covers."""
+    covers = []
+    devices = config.get(CONF_COVERS)
+
+    for device_id, device_config in devices.items():
+        api_key = retrieve_api_key(device_config.get(CONF_USERNAME),
+                                   device_config.get(CONF_PASSWORD),
+                                   device_config.get(CONF_DEVICE, device_id))
+        device_id_valid = check_device_id(device_config.get(CONF_USERNAME),
+                                          device_config.get(CONF_PASSWORD),
+                                          device_config.get(CONF_DEVICE,
+                                                            device_id))
+        if api_key is not None and device_id_valid:
+            args = {
+                'name': "ryobigdo_" + device_config.get(CONF_DEVICE,
+                                                        device_id),
+                'device_id': device_config.get(CONF_DEVICE, device_id),
+                'username': device_config.get(CONF_USERNAME),
+                'password': device_config.get(CONF_PASSWORD),
+                'api_key': api_key
+            }
+            covers.append(RyobiCover(hass, args,
+                                     supported_features=(SUPPORT_OPEN |
+                                                         SUPPORT_CLOSE)))
+
+    add_devices(covers)
+
+
+def retrieve_api_key(username, password, device_id):
+    """Getting api_key from Ryobi."""
+    _LOGGER.debug("Getting api_key from Ryobi")
+    req = requests.post(RYOBI_API_KEY_URL, params={
+        'username': username,
+        'password': password
+    })
+    _LOGGER.debug(req.status_code)
+    if req.status_code == 200:
+        _LOGGER.debug("auth OK. api_key retrieved")
+        req_meta = req.json()['result']['metaData']
+        local_api_key = req_meta['wskAuthAttempts'][0]['apiKey']
```

Comment:
This is too much protocol specific information. Please extract this into a 3rd party lib.